enclave/core/tracee.c

```diff
@@ -69,6 +77,63 @@ void oe_log_init_ecall(const char* enclave_path, uint32_t log_level)
     }
 }
 
+static void* oe_enclave_log_context = NULL;
+static oe_enclave_log_callback_t oe_enclave_log_callback = NULL;
+
+oe_result_t oe_enclave_log_set_callback(
+    void* context,
+    oe_enclave_log_callback_t callback)
+{
+    oe_result_t result = OE_OK;
+
+    OE_CHECK_NO_TRACE(oe_mutex_lock(&_log_lock));
+
+    oe_enclave_log_context = context;
+    oe_enclave_log_callback = callback;
+    oe_mutex_unlock(&_log_lock);
+
+    OE_TRACE_INFO("enclave logging callback is set");
+
+done:
+    return result;
+}
+
+static oe_result_t oe_log_enclave_message(
+    oe_log_level_t level,
+    const char* message)
+{
+    oe_result_t result = OE_FAILURE;
+
+    if (!message)
+        OE_RAISE_NO_TRACE(OE_UNEXPECTED);
+
+    // Take the log file lock.
+    OE_CHECK_NO_TRACE(oe_mutex_lock(&_log_lock));
+
+    if (oe_enclave_log_callback)
+    {
+        (oe_enclave_log_callback)(
+            oe_enclave_log_context, level, (uint64_t)oe_thread_self(), message);
+    }
+    else
+    {
+        // Check if this message should be skipped
+        if (level > _active_log_level)
+        {
+            result = OE_OK;
+            goto done;
+        }
+
+        OE_CHECK_NO_TRACE(oe_log_ocall(level, message));
+    }
+
+    result = OE_OK;
+
+done:
+    oe_mutex_unlock(&_log_lock);
```

Comment:
. If !message you will go to done without locking the mutex. can you verify that if the mutex is not locked, that this does not cause any issues?